### PR TITLE
[macOS] Implement hit testing and handle platform view cursor changes

### PR DESCRIPTION
**Pull request by knopp**
Fixes https://github.com/flutter/flutter/issues/129085

## Changes to `FlutterMouseCursorPlugin`

- `none` cursor is not handled by hiding the cursor anymore. That was too big of a hammer as it also affects other views (and platform views). Instead empty cursor is created from empty `NSImage`.

- Cursor plugin now notifies the engine when cursor has changed. The engine forwards the change to last `FlutterView` that handled mouse event. This is necessary because on occasion `FlutterView` needs to be able to restore cursor without framework being involved.

## Preventing PlatformView from changing cursor when it is obscured by Flutter Content.

Generally in Cocoa cursor changes are done as response to `mouseMoved` event, which is driven by a `NSTrackingArea`. The issue here is that this is not affected by hit testing and tracking areas form a hierarchy parallel to view hierarchy and are not affected by being obscured by another view (or tracking area). This means that platform view will receive `mouseMoved` event even when is obscured by Flutter content. To work around this, the mutator view puts a tracking area above platform view, which means it gets the mouseMove event first, and when it decides that mouse is over Flutter content, it will prevent platform view from changing the cursor for the rest of RunLoop turn (see `NSCursor+IgnoreChange`).

## Actual hit testing

This part is rather straightforward, the area where FlutterContent obscures mutator view is provided to the mutator view, which will return `nil` from `hitTest:` when the point is in the obscured area.

## Example of hit testing

https://github.com/flutter/engine/assets/96958/bbac0cfd-8c44-44d3-addd-921c91a8a539

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Ping @cbracken

**Review by cbracken — COMMENTED** [review]
Looking at the embedder side changes right now, but a few comments on the pointer event/cursor side first.

File: shell/platform/darwin/macos/framework/Source/NSCursor+IgnoreChange.mm [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  method_exchangeImplementations(class_getInstanceMethod(originalClass, originalSelector),
+                                 class_getInstanceMethod(originalClass, replacementSelector));
+}
+
++ (void)load {
+  [self replaceSelector:@selector(set)
+              fromClass:[NSCursor class]
+           withSelector:@selector(_flutterSetCursorOverride)
+              fromClass:[NSCursor class]];
```
[the remark this answers is not in the document]
Comment (knopp) [reply shell/platform/darwin/macos/framework/Source/NSCursor+IgnoreChange.mm:47]: Swizzling in general does not prevent getting into AppStore. Using private / undocumented API does, which is not the case here. This also should not interfere with the rest of application, it just prevents changing cursor until the rest of RunLoop tick when Flutter content is over native view and already handled cursor change.

The application will be working without it, but with cursor flicker when moving over Flutter contents that have native view with active `NSTrackingArea` underneath (i.e. `NSTextField`).

**Comment by gspencergoog** [comment]
@knopp Are you still planning on working on this?  If not, could you either close it or mark it as draft so we can get it off the PR backlog? (Desktop Triage)

**Comment by cbracken** [comment]
This is pending another round of review from me, but been stuck working on two internal customer issues, I'll try to get back to this soon. Apologies for the delay.

**Comment by gspencergoog** [comment]
Converting this to Draft, as @cbracken is still looking at alternatives.

**Comment by goderbauer** [comment]
Spoke to @cbracken and he says this is still on his radar.

**Comment by knopp** [comment]
I have removed the method swizzling part. The downside is slightly reduced fidelity in a corner case where platform view has a NSTrackingArea that it uses to change mouse cursor and is overlaid by Flutter content. Given that method swizzling seems a bit contentious I think this is a reasonable tradeoff.

File: shell/platform/darwin/macos/framework/Source/FlutterCompositor.mm [hunk 2]
```diff
@@ -3,22 +3,30 @@
...
+    const auto& layer = layers[i];
+    if (layer->type == kFlutterLayerContentTypePlatformView) {
+      layers_copy.push_back(PlatformViewLayer(layer));
+    } else if (layer->type == kFlutterLayerContentTypeBackingStore) {
+      std::vector<FlutterRect> rects;
+      auto present_info = layer->backing_store_present_info;
+      if (present_info) {
+        for (size_t j = 0; j < present_info->paint_region->rects_count; j++) {
+          rects.push_back(present_info->paint_region->rects[j]);
+        }
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterCompositor.mm:24]: Consider pre-reserving size of rects, and using std::copy which might have better performance:
```cc
rects.reserve(present_info->paint_region->rects_count);
std::copy(present_info->paint_region->rects,
          present_info->paint_region->rects + present_info->paint_region->rects_count,
          back_inserter(rects));
```

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

**Pull request merged**